src/picongpu/include/fields/background/templates/bgrTWTS.hpp

```diff
@@ -0,0 +1,953 @@
+                 * transversally centered with respect to the global simulation volume. */
+                PMacc::math::Vector<float3_X, FieldE::numComponents> eFieldPositions = 
+                                    dummyCast(fieldSolver::NumericalCellType::getEFieldPosition());
+                
+                PMacc::math::Vector<float3_64,FieldE::numComponents> eFieldPositions_SI;
+                
+                for( uint32_t i = 0; i < FieldE::numComponents; ++i ) // cellIdx Ex, Ey and Ez
+                {
+                    eFieldPositions[i]   += ( dummyCast( floatD_X(cellIdx) ) - laserOrigin );
+                    eFieldPositions_SI[i] = precisionCast<float_64>(eFieldPositions[i])
+                                            * cellDimensions;
+                    
+                    /*  Since, the laser propagation direction encloses an angle of phi with the
+                     *  simulation y-axis (i.e. direction of sliding window), the positions vectors
+                     *  are rotated around the simulation x-axis before calling the TWTS field
+                     *  functions. Note: The TWTS field functions are in non-rotated frame and only
+                     *  use the angle phi to determine the required amount of pulse front tilt.
+                     *  RotationMatrix[PI/2+phi].(y,z) (180Deg-flip at phi=90Deg since coordinate
+                     *  system in paper is oriented the other way round.) */
+                    eFieldPositions_SI[i] = float3_64( (eFieldPositions_SI[i]).x(),
+                       -sin(phi)*(eFieldPositions_SI[i]).y()-cos(phi)*(eFieldPositions_SI[i]).z(),
+                       +cos(phi)*(eFieldPositions_SI[i]).y()-sin(phi)*(eFieldPositions_SI[i]).z() );
+                }
+             
+            return eFieldPositions_SI;
+        }
+         
+        template<>
+        HDINLINE PMacc::math::Vector<float3_64,FieldE::numComponents>
+        TWTSFieldE::getEfieldPositions_SI<DIM2>(const DataSpace<simDim>& cellIdx) const
+        {
+            const float2_64 cellDimensions = ( float_64( picongpu::cellSize.x() ),
+                                               float_64( picongpu::cellSize.y() ) )
+                                             * unit_length;
+            
+            /* TWTS laser coordinate origin is centered transversally and defined longitudinally by
+               the laser center in y (usually maximum of intensity). */
+            float2_X laserOrigin = float2_X( halfSimSize.x(), focus_y_SI/cellDimensions.y() );
+            
+            /* For the Yee-Cell shifted fields, obtain the fractional cell index components and add
+             * that to the total cell indices. The physical field coordinate origin is transversally
+             * centered with respect to the global simulation volume. */
+            PMacc::math::Vector<float3_X, FieldE::numComponents> eFieldPositions =
```

Comment:
This would not work, because then eFieldPositions would be a vector of 2D vectors in the 2D case. It should be 3D, since I am always calling the calcTWTSEx(), which requires a 3D position vector. Since I have to rotate the coordinate system and set one coordinate to zero (except potentially for a intraCell-shift), I have to differentiate between the coordinates beyond shuffling components. Furthermore without pre-compiler directives all code needs to parse for all dimensions. Here I avoid this dilemma by providing a way to cast 2D vectors to 3D vectors via `dummyCast`. Better solutions are welcome.